In [ ]:
!pip install mlxtend

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
toy_dataset = [['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],

        ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],

        ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],

        ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],

      ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]


In [ ]:
##Apriori Preparation

import mlxtend
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
te=TransactionEncoder()
te_ary=te.fit(toy_dataset).transform(toy_dataset)    #Apply one-hot-encoding on our dataset
toys=pd.DataFrame(te_ary, columns=te.columns_)  #Creating a new DataFrame from our Numpy array
toys

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


Dress  Earrings    Hat  Heels  Necklace  Pants  Sandals  Scarf  Shirt  \
0  False     False   True  False     False   True    False   True  False   
1  False     False   True  False     False   True    False  False  False   
2   True     False  False   True     False   True     True   True  False   
3   True      True   True   True      True  False    False   True  False   
4  False      True  False  False     False   True    False   True   True   

   Skirt  Sneakers  Sunglasses  
0   True      True       False  
1   True      True        True  
2  False     False       False  
3  False     False       False  
4   True     False       False

In [ ]:
##Support Code
from mlxtend.frequent_patterns import apriori
##Support with column names
frequent_itemsets=apriori(toys, min_support=0.6, use_colnames=True)
#Instead of column indices we can use column names.
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support        itemsets
0      0.6           (Hat)
1      0.8         (Pants)
2      0.8         (Scarf)
3      0.6         (Skirt)
4      0.6  (Scarf, Pants)
5      0.6  (Pants, Skirt)

In [ ]:
##Confidence Code
##In case we want to extract rules based on other metrics like confidence,
##we can use association_rules from mlxtend.frequent_patterns library.

from mlxtend.frequent_patterns import association_rules
association_rules(frequent_itemsets,metric="confidence",min_threshold=0.7) # associate items


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents consequents  antecedent support  consequent support  support  \
0     (Scarf)     (Pants)                 0.8                 0.8      0.6   
1     (Pants)     (Scarf)                 0.8                 0.8      0.6   
2     (Pants)     (Skirt)                 0.8                 0.6      0.6   
3     (Skirt)     (Pants)                 0.6                 0.8      0.6   

   confidence    lift  leverage  conviction  zhangs_metric  
0        0.75  0.9375     -0.04         0.8          -0.25  
1        0.75  0.9375     -0.04         0.8          -0.25  
2        0.75  1.2500      0.12         1.6           1.00  
3        1.00  1.2500      0.12         inf           0.50

In [ ]:
##Lift code
##Associating based on Lift

from mlxtend.frequent_patterns import association_rules
association_rules(frequent_itemsets,metric="lift",min_threshold=1.25)


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents consequents  antecedent support  consequent support  support  \
0     (Skirt)     (Pants)                 0.6                 0.8      0.6   

   confidence  lift  leverage  conviction  zhangs_metric  
0         1.0  1.25      0.12         inf            0.5

## Read in the Data

In [ ]:
# Read in the data
import pandas as pd
Market = pd.read_csv("/content/Market_Basket_Optimisation.csv", header=None)

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Convert the DataFrame into a list of lists and dropping nans
transactions = []
for i in range(0,len(Market)):
    transactions.append([str(Market.values[i, j]) for j in range(0, Market.shape[1]) if pd.notna(Market.values[i, j])])

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
# Apriori Preparation
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)  # Apply one-hot-encoding on our dataset
SuperMarket = pd.DataFrame(te_ary, columns=te.columns_)  # Creating a new DataFrame from our Numpy array

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


In [ ]:
##Support Code
from mlxtend.frequent_patterns import apriori
##Support with column names
frequent_itemsets=apriori(SuperMarket, min_support=0.05, use_colnames=True)
#Instead of column indices we can use column names.
frequent_itemsets

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


support                    itemsets
0   0.087188                   (burgers)
1   0.081056                      (cake)
2   0.059992                   (chicken)
3   0.163845                 (chocolate)
4   0.080389                   (cookies)
5   0.051060               (cooking oil)
6   0.179709                      (eggs)
7   0.079323                  (escalope)
8   0.170911              (french fries)
9   0.063325           (frozen smoothie)
10  0.095321         (frozen vegetables)
11  0.052393             (grated cheese)
12  0.132116                 (green tea)
13  0.098254               (ground beef)
14  0.076523            (low fat yogurt)
15  0.129583                      (milk)
16  0.238368             (mineral water)
17  0.065858                 (olive oil)
18  0.095054                  (pancakes)
19  0.071457                    (shrimp)
20  0.050527                      (soup)
21  0.174110                 (spaghetti)
22  0.068391                  (tomatoes)
23  0.062525                    (turkey)
24  0.058526          (whole wheat rice)
25  0.052660  (chocolate, mineral water)
26  0.050927       (mineral water, eggs)
27  0.059725  (spaghetti, mineral water)

In [ ]:
##Confidence Code
from mlxtend.frequent_patterns import association_rules
rules=association_rules(frequent_itemsets,metric="confidence",min_threshold=0.2) # associate items
rules


/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents      consequents  antecedent support  consequent support  \
0      (chocolate)  (mineral water)            0.163845            0.238368   
1  (mineral water)      (chocolate)            0.238368            0.163845   
2  (mineral water)           (eggs)            0.238368            0.179709   
3           (eggs)  (mineral water)            0.179709            0.238368   
4      (spaghetti)  (mineral water)            0.174110            0.238368   
5  (mineral water)      (spaghetti)            0.238368            0.174110   

    support  confidence      lift  leverage  conviction  zhangs_metric  
0  0.052660    0.321400  1.348332  0.013604    1.122357       0.308965  
1  0.052660    0.220917  1.348332  0.013604    1.073256       0.339197  
2  0.050927    0.213647  1.188845  0.008090    1.043158       0.208562  
3  0.050927    0.283383  1.188845  0.008090    1.062815       0.193648  
4  0.059725    0.343032  1.439085  0.018223    1.159314       0.369437  
5  0.059725    0.250559  1.439085  0.018223    1.102008       0.400606

In [ ]:
##Lift code
##Associating based on Lift

from mlxtend.frequent_patterns import association_rules
rules=association_rules(frequent_itemsets,metric="lift",min_threshold=1.15)
rules

/usr/local/lib/python3.10/dist-packages/ipykernel/ipkernel.py:283: DeprecationWarning: `should_run_async` will not call `transform_cell` automatically in the future. Please pass the result to `transformed_cell` argument and any exception that happen during thetransform in `preprocessing_exc_tuple` in IPython 7.17 and above.
  and should_run_async(code)


antecedents      consequents  antecedent support  consequent support  \
0      (chocolate)  (mineral water)            0.163845            0.238368   
1  (mineral water)      (chocolate)            0.238368            0.163845   
2  (mineral water)           (eggs)            0.238368            0.179709   
3           (eggs)  (mineral water)            0.179709            0.238368   
4      (spaghetti)  (mineral water)            0.174110            0.238368   
5  (mineral water)      (spaghetti)            0.238368            0.174110   

    support  confidence      lift  leverage  conviction  zhangs_metric  
0  0.052660    0.321400  1.348332  0.013604    1.122357       0.308965  
1  0.052660    0.220917  1.348332  0.013604    1.073256       0.339197  
2  0.050927    0.213647  1.188845  0.008090    1.043158       0.208562  
3  0.050927    0.283383  1.188845  0.008090    1.062815       0.193648  
4  0.059725    0.343032  1.439085  0.018223    1.159314       0.369437  
5  0.059725    0.250559  1.439085  0.018223    1.102008       0.400606

**Interpretation : **

•(chocolate) -> (mineral water)

-Support (chocolate): Approximately 16.38% of all transactions contain chocolate.

-Support (mineral water): Roughly 23.84% of all transactions contain mineral water.

-Support (chocolate and mineral water): About 5.27% of all transactions contain both chocolate and mineral water.

-Confidence: When a customer buys chocolate, there's a 32.14% chance they will also buy mineral water.

-Lift: The purchase of chocolate is 1.35 times more likely when mineral water is also purchased compared to the purchase of chocolate alone.

►Interpretation: Customers who buy chocolate are 32.14% likely to also buy mineral water. This suggests a moderate association between these two items. It might be useful for marketing or store layout decisions.

•(spaghetti) -> (mineral water)

-Support (spaghetti): Around 17.41% of all transactions contain spaghetti.

-Support (mineral water): Approximately 23.84% of all transactions contain mineral water.

-Support (spaghetti and mineral water): About 5.97% of all transactions contain both spaghetti and mineral water.

-Confidence: When a customer buys spaghetti, there's a 34.30% chance they will also buy mineral water.

-Lift: The purchase of spaghetti is 1.43 times more likely when mineral water is also purchased compared to the purchase of spaghetti alone.

►Interpretation: Customers who buy spaghetti are 34.30% likely to also buy mineral water. This suggests a moderate association between these two items, which could be valuable for product placement or promotions.




